torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -597,6 +774,232 @@ static void PrepareForRemoveMutations(MutationRemover& mr, Block* b) {
+//   ...
+//   %B = prim::GetAttr[name="B"](%A)
+//   ...
+//   %weight = prim::GetAttr[name="scale"](%B)
+//   ...
+std::deque<std::string> findSubModuleAttr(
+    Value* input,
+    std::string& name,
+    Module& attrModule,
+    const std::shared_ptr<Graph>& graph) {
+  Node* node = input->node();
+  std::deque<std::string> moduleNames;
+
+  // Loop starts from inner submodule and follows the chain until reaches the
+  // top module.
+
+  auto selfNode = graph->nodes().begin();
+  auto n = *selfNode;
+  while (node->outputs().at(0)->type() != n->output()->type()) {
+    if (node->kind() == prim::GetAttr) {
+      moduleNames.push_front(node->s(attr::name));
+      node = node->inputs()[0]->node();
+    } else {
+      return moduleNames;
+    }
+  }
+  // Assign the inner module to attrModule.
+  for (auto& moduleName : moduleNames) {
+    attrModule = attrModule.attr(moduleName).toModule();
+  }
+  return moduleNames;
+}
+
+Node* insertCloneBeforeNode(
+    std::shared_ptr<Graph> graph,
+    Value* orig_data,
+    Node* node) {
+  Node* newNode = nullptr;
+  if (orig_data->type()->kind() == TypeKind::ListType) {
+    // Create an aten::list to clone the list in graph inputs
+    newNode = graph->create(aten::list, /*num_outputs =*/1);
+    newNode->addInput(orig_data);
+    newNode->output()->setType(orig_data->type());
+    newNode->insertBefore(node);
+  } else if (orig_data->type()->kind() == TypeKind::TensorType) {
+    auto* noneNode = graph->create(prim::Constant);
+    noneNode->output()->setType(NoneType::get());
+    newNode = graph->create(aten::clone, /*num_outputs =*/1);
+    newNode->addInput(orig_data);
+
+    newNode->addInput(noneNode->output());
+    newNode->output()->setType(orig_data->type());
+    newNode->insertBefore(node);
+    noneNode->insertBefore(newNode);
+  } // TODO: Handle float/int attributes
+  return newNode;
+}
+
+Value* registerSetAttrInBlocks(
+    const std::shared_ptr<Graph>& graph,
+    Block* block,
+    Value* newValue,
+    Value* origValue,
+    std::string output_name) {
+  auto cloneNode = insertCloneBeforeNode(graph, newValue, block->return_node());
+  auto next_node = block->owningNode();
+
+  RegisterInplaceNodeInIfBlocks(
+      origValue, cloneNode->output(), block, next_node, output_name);
+
+  return MatchIfBlocksOutputForValue(origValue, block, cloneNode->output());
+}
+
+Value* findArgumentAsInputParam(
+    const std::shared_ptr<Graph>& graph,
+    std::string& name,
+    IValue& attr) {
+  for (auto input : graph->inputs()) {
+    if ((input->debugName() == name))
+      return input;
+  }
+  throw std::runtime_error(
+      "Attribute is not part of model parameters. Cannot handle SetAttr and GetAttr nodes for : " +
+      name);
+}
+
+std::unordered_map<std::string, Value*> registerInplaceOpAsBlockOutputs(
+    MutationRemover& mr,
+    Block* block,
+    const std::shared_ptr<Graph>& graph,
+    Module module_,
+    std::unordered_map<std::string, Value*> setAttrValues) {
```

Comment:
why pass by value argument?

Reply:
Sorry, what is the question?

Reply:
wondering if setAttrValues could be passed by reference